lib/eth/address.rb

```diff
@@ -0,0 +1,85 @@
+# Copyright (c) 2016-2022 The Ruby-Eth Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Eth
+  
+  class Address
+
+    # Constructor of the `Eth::Address` class. Creates a new hex
+    # prefixed address
+    #
+    # @param address [String] hex string representing an ethereum address.
+    def initialize address
+      @address = Utils.prefix_hex address
+    end
+
+    # Check that the address is valid
+    #
+    # @return [Bool]
+    def valid?
+      if !matches_any_format?
+        false
+      elsif not_checksummed?
+        true
+      else
+        checksum_matches?
+      end
+    end
+
+    # Generate a checksummed address
+    #
+    # @return address [String] prefixed hex string representing an checksummed address.
+    def checksummed
+      raise "Invalid address: #{address}" unless matches_any_format?
+
+      cased = unprefixed.chars.zip(checksum.chars).map do |char, check|
+        check.match(/[0-7]/) ? char.downcase : char.upcase
+      end
+
+      Utils.prefix_hex(cased.join)
+    end
+
+    private
```

Comment:
Agreed. This entire class needs more testing.

Reply:
@q9f I added tests for the `Address` class but only tested the private methods indirectly. Otherwise we will need to unit test the methods with Ruby's `send` method: 

```
it "tests the private method directly" do
  expect(MyClass.send(:my_private_method)).to eq("private method")
end
```

with feels kind of clunky and breaks encapsulation.

What do you think?

Reply:
This is fine. :+1: